### PR TITLE
Support multiple rates in monthly SMS usage

### PR DESCRIPTION
**Pull request by benthorner**
https://www.pivotaltracker.com/story/show/181935935

This makes use of the new API fields in [^1] to compute
the SMS and letter costs, as well as the free/paid split
of SMS with the free allowance.

With this in place, it's then a simple matter to switch to 
support multiple monthly rates.

I found the original code quite hard to navigate, so the
first few commits are a bit of refactoring to make the main
changes easier to make and follow.

[^1]: https://github.com/alphagov/notifications-api/pull/3520

### CODE REVIEW

**Comment by idavidmcdonald** [comment]
Unfortunately I don't have time to look at this PR properly but I will leave hopefully a helpful comment.

We current support SMS rate changes on the 1st day of the month. This PR would add support for SMS rate changes on other days of the month.

I'm not sure whether we should support this because [you aren't gonna need it](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it). I don't think we have any evidence that we need this yet. We do expect another rate change but the impression I've got is that we will likely try and do it on the 1st of the month again. I'm cautious about adding functionality for something we don't currently know we will need.

Saying that, at a very quick glance there is likely some very good refactoring in here which are useful regardless. Equally, it may be the case that the change to support sms rate changes mid month is so small/easy/not complex or even reduces complexity of the current code that it is worth merging regardless too.

I'd encourage you to get some steer on whether this is a useful feature from Katie and Caley if you haven't already.

Even if the answer is that we probably aren't going to need this but it is still more beneficial to have in our codebase than not, that is OK too but worth being explicit about that I think please.

**Review by idavidmcdonald — APPROVED** [review]
Well this is a fairly complex area of the codebase but overall I think there are some useful changes in here which help with readability.

Nice work for managing to get to grips with all this stuff! And good commit messages as always.

Two very minor comments for you...

File: tests/app/main/views/test_dashboard.py [hunk 1]
```diff
@@ -1093,87 +1093,21 @@ def test_usage_page_monthly_breakdown(
     assert '1,230 text messages at 1.70p' in monthly_breakdown
 
 
-@freeze_time("2012-04-30 12:12:12")
-def test_usage_page_displays_letters_ordered_by_postage(
-    mocker,
+@freeze_time("2012-03-31 12:12:12")
+def test_usage_page_letter_breakdown_ordering(
```
[the remark this answers is not in the document]
Comment (benthorner) [reply tests/app/main/views/test_dashboard.py]: Sure, I've changed in https://github.com/alphagov/notifications-admin/pull/4229/commits/bb0fb73bc82a4e83712185fe7eb5ae8ffb7205b8 to clarify it's by postage and rate, tweaking the test data so we at least attempt to show the rate-based sorting - not bullet proof but slightly more coverage than we had before.

Would you like to take another look?

[the remark this answers is not in the document]
Comment (idavidmcdonald) [reply tests/app/main/views/test_dashboard.py]: All good, thanks. 🚢

**Pull request merged**